src/common/data/fields.d.ts

```diff
@@ -1 +1,372 @@
+  type: typeof Number;
+  required: true;
+  default: 0;
+  validate: typeof Number.isInteger;
+  validationError: 'The provided {name} {field} field does not have an integer value';
+}
+
+/**
+ * A field used for integer sorting of a Document relative to its siblings
+ */
+export declare const INTEGER_SORT_FIELD: IntegerSortField;
+
+interface TimestampField extends DocumentField<number> {
+  type: typeof Number;
+  required: false;
+  default: typeof Date.now;
+  nullable: false;
+}
+
+/**
+ * A numeric timestamp field which may be used in a Document.
+ */
+export declare const TIMESTAMP_FIELD: TimestampField;
+
+/**
+ * Validate that the ID of a Document object is either null (not yet saved) or a valid string.
+ * @param id - The _id to test
+ * @returns Is it valid?
+ */
+declare function _validateId(id: string | null): boolean;
+
+interface DocumentId extends DocumentField<string | null> {
+  type: typeof String;
+  required: true;
+  default: null;
+  nullable: false;
+  validate: typeof _validateId;
+  validationError: 'Invalid value provided for {name} id';
+}
+
+/**
+ * The standard identifier for a Document.
+ */
+export declare const DOCUMENT_ID: DocumentId;
+
+interface DocumentPermissions extends DocumentField<Partial<Record<string, ValueOf<typeof CONST.ENTITY_PERMISSIONS>>>> {
+  type: typeof Object;
+  required: true;
+  nullable: false;
+  default: { default: typeof CONST.ENTITY_PERMISSIONS.NONE };
+  validate: typeof _validatePermissions;
+  validationError: 'Invalid data structure encountered for {name} permissions';
+}
+
+/**
+ * The standard permissions object which may be included by a Document.
+ */
+export declare const DOCUMENT_PERMISSIONS: DocumentPermissions;
+
+/**
+ * Validate the structure of the permissions object: all anykeys are valid IDs and all values are permission levels
+ * @param perms - The provided permissions object
+ * @returns Is the object valid?
+ */
+declare function _validatePermissions(perms: object): boolean;
+
+interface ForeignDocumentField<
+  T extends {
+    type: {
+      readonly documentName: string;
+    };
+    required?: boolean;
+    nullable?: boolean;
+    default?: any;
+  }
+> extends DocumentField<string | null> {
+  type: typeof String;
+  required: T extends {
+    required: true;
+  }
+    ? true
+    : false;
+  nullable: T extends {
+    nullable?: true;
+  }
+    ? true
+    : T extends {
+        nullable: false;
+      }
+    ? false
+    : boolean;
+  default: T extends {
+    default: infer U;
+  }
+    ? U
+    : null;
+  clean: (d: unknown) => string | null;
+  validate: typeof _validateId;
+  validationError: `Invalid {name} {field} which must be a ${T['type']['documentName']} document id`;
+}
+
+/**
+ * Create a foreign key field which references a primary Document id
+ */
+export declare function foreignDocumentField<
+  T extends { type: { readonly documentName: string }; required?: boolean; nullable?: boolean; default?: any }
```

Comment:
I think the type constraint of this generic should be put into an interface, that way we can use for both the function and the interface.

Reply:
I added that.